### DISCUSSION TITLE
Theme: CursorKind bar is using system terminal font color (can be effectively invisible)

### DISCUSSION THREAD

**Discussion by nuke-web3** [helix-editor/helix#8234] [Troubleshooting]
### Summary

with any `bar` selection for a cursor, most evidently on the opposite of the _OS/system_ theme, the bar is illegible: 
```toml
[editor.cursor-shape]
insert = "bar"
```
Example in normal mode (catppuccin_latte):

![image](https://github.com/helix-editor/helix/assets/35669742/2903b61d-ebf0-4a82-8c62-ac07547685cd)

insert mode (cursor just past `as`:

![image](https://github.com/helix-editor/helix/assets/35669742/043b94a8-443e-4c6b-934f-f30dffccbeb2)

### Reproduction Steps

Try out just about any theme, us a bar cursor, and I _think_ it is using the underlying _system color_ for terminal font to render, not some color from the theme...? In my case, wezterm using 'Monokai Dark (Gogh)'

### Platform

Linux

### Terminal Emulator

wezterm 20230712-072601-f4abf8fd

### Helix Version

helix 23.05 (7f5940be)

**Comment by nuke-web3** [comment]
Confirmed - system term font color is used: 
 
![image](https://github.com/helix-editor/helix/assets/35669742/da538a05-3da8-4976-b887-3d3a578a03b9)

**Comment by the-mikedavis** [comment]
AFAIK there isn't a way for a terminal application to tell the terminal which color to use for the cursor. For block cursors we actually hide the terminal's cursor and highlight the cell at the cursor's position with the color from the theme. There isn't a way to highlight a part of a cell so we can't use the same trick for the bar cursor.

**Reply by nuke-web3** [reply]
I did a bit of poking, it seems that there _may_ be a way based on nvim having a similar problem before:
- https://github.com/alacritty/alacritty/issues/826#issuecomment-488705325
- https://github.com/neovim/neovim/pull/7277
- https://github.com/mauke/unibilium/blob/master/doc/unibi_var_from_str.pod

Based on this perhaps https://github.com/helix-editor/helix/issues/5319 is related to solving this... something to do with TERMINFO access?

**Reply by nuke-web3** [reply]
If using the _system_ cursor isn't an option, is a hack possible to emulate the block cursor behavior with something that is _close_ to a bar? Perhaps drawing over the cell with a unicode character that is "close enough" `U+23C8` from https://en.wikipedia.org/wiki/Miscellaneous_Technical looks like perhaps one that could do that

**Reply by the-mikedavis** [reply]
Ah it looks like there are escape codes for setting this: https://github.com/wez/wezterm/blob/70931f58f1c7a57fd377f9862597e7b1fe070b52/docs/escape-sequences.md?plain=1#L396-L398 / https://github.com/alacritty/alacritty/blob/e29da339b4960b056b349f6bda172ca70d3c16a0/src/ansi.rs#L792-L803. I'm not sure how standard that sequence is but it seems like it has buy-in from popular terminal emulators. We could probably make use of that but first it would need to be contributed upstream to the terminal manipulation library we use: https://github.com/crossterm-rs/crossterm.

We can't use a unicode bar to do the same trick because we can't combine the bar with a character in that cell. So if there were any content in the same cell as the cursor we would need to choose between drawing the cursor and the content.

**Comment by getreu** [comment]
After setting 

```toml
[editor.cursor-shape]
insert = "bar"
```

the theme `gruber-darker` has the same problem, while the theme `autumn` hasn't. I compared `autumn.toml` and `gruber-darker.toml` but I did not find out what difference causes the disappearance of the cursor.

BTW: I think this should be considered being a bug.

**Comment by getreu** [comment]
#9002 is related
